src/VirtoCommerce.Platform.Core/ChangesUtils/ChangesDetector.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.Exceptions;
+
+namespace VirtoCommerce.Platform.Core.ChangesUtils
+{
+    /// <summary>
+    /// Object changes gathering utility.
+    /// </summary>
+    public static class ChangesDetector
+    {
+        private static readonly ConcurrentDictionary<Type, IEnumerable<ChangesDetectorPropertyInfo>> _changesDetectorPropertyInfosCache = new ConcurrentDictionary<Type, IEnumerable<ChangesDetectorPropertyInfo>>();
+
+        /// <summary>
+        /// Gather changes info for objects by properties, defined in type T.
+        /// Properties to compare should be annotated with <see cref="UseInChangesDetectorAttribute"/>.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="newObj">An object</param>
+        /// <param name="oldObj">An object compare to</param>
+        /// <returns></returns>
+        public static ListDictionary<string, string> GatherByType<T>(T newObj, T oldObj)
+        {
+            var objType = typeof(T);
+            return GatherByType(objType, newObj, oldObj);
+        }
+
+        /// <summary>
+        /// Gather changes info for objects by properties, defined in type objType.
+        /// Properties to compare should be annotated with <see cref="UseInChangesDetectorAttribute"/>.
+        /// </summary>
+        /// <param name="objType">Type which properties should be used to compare</param>
+        /// <param name="newObj">An object</param>
+        /// <param name="oldObj">An object compare to</param>
+        /// <returns></returns>
+        public static ListDictionary<string, string> GatherByType(Type objType, object newObj, object oldObj)
```

Comment:
I think we should make them generic too, to prevent problems in getters when `newObj` and `oldObj` have different types.